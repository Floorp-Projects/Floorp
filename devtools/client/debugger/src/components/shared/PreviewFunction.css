/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.function-signature {
  align-self: center;
}

.function-signature .function-name {
  color: var(--theme-highlight-blue);
}

.function-signature .param {
  color: var(--theme-highlight-red);
}

.function-signature .paren {
  color: var(--object-color);
}

.function-signature .comma {
  color: var(--object-color);
}
