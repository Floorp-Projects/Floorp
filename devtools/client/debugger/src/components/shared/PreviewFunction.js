/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import PropTypes from "prop-types";

import { times, zip, flatten } from "lodash";

import { formatDisplayName } from "../../utils/pause/frames";

import "./PreviewFunction.css";

const IGNORED_SOURCE_URLS = ["debugger eval code"];

export default class PreviewFunction extends Component {
  renderFunctionName(func) {
    const { l10n } = this.context;
    const name = formatDisplayName(func, undefined, l10n);
    return <span className="function-name">{name}</span>;
  }

  renderParams(func) {
    const { parameterNames = [] } = func;
    const params = parameterNames.filter(Boolean).map(param => (
      <span className="param" key={param}>
        {param}
      </span>
    ));

    const commas = times(params.length - 1).map((_, i) => (
      <span className="delimiter" key={i}>
        {", "}
      </span>
    ));

    return flatten(zip(params, commas));
  }

  jumpToDefinitionButton(func) {
    const { location } = func;

    if (
      location &&
      location.url &&
      !IGNORED_SOURCE_URLS.includes(location.url)
    ) {
      const lastIndex = location.url.lastIndexOf("/");

      return (
        <button
          className="jump-definition"
          draggable="false"
          title={`${location.url.slice(lastIndex + 1)}:${location.line}`}
        />
      );
    }
  }

  render() {
    const { func } = this.props;
    return (
      <span className="function-signature">
        {this.renderFunctionName(func)}
        <span className="paren">(</span>
        {this.renderParams(func)}
        <span className="paren">)</span>
        {this.jumpToDefinitionButton(func)}
      </span>
    );
  }
}

PreviewFunction.contextTypes = { l10n: PropTypes.object };
