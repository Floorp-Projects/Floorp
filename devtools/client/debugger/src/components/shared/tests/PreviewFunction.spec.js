/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";
import PreviewFunction from "../PreviewFunction";

function render(props) {
  return shallow(<PreviewFunction {...props} />, { context: { l10n: L10N } });
}

describe("PreviewFunction", () => {
  it("should return a span", () => {
    const item = { name: "" };
    const returnedSpan = render({ func: item });
    expect(returnedSpan).toMatchSnapshot();
    expect(returnedSpan.name()).toEqual("span");
  });

  it('should return a span with a class of "function-signature"', () => {
    const item = { name: "" };
    const returnedSpan = render({ func: item });
    expect(returnedSpan.hasClass("function-signature")).toBe(true);
  });

  it("should return a span with 3 children", () => {
    const item = { name: "" };
    const returnedSpan = render({ func: item });
    expect(returnedSpan.children()).toHaveLength(3);
  });

  describe("function name", () => {
    it("should be a span", () => {
      const item = { name: "" };
      const returnedSpan = render({ func: item });
      expect(
        returnedSpan
          .children()
          .first()
          .name()
      ).toEqual("span");
    });

    it('should have a "function-name" class', () => {
      const item = { name: "" };
      const returnedSpan = render({ func: item });
      expect(
        returnedSpan
          .children()
          .first()
          .hasClass("function-name")
      ).toBe(true);
    });

    it("should be be set to userDisplayName if defined", () => {
      const item = {
        name: "",
        userDisplayName: "chuck",
        displayName: "norris",
      };
      const returnedSpan = render({ func: item });
      expect(
        returnedSpan
          .children()
          .first()
          .first()
          .text()
      ).toEqual("chuck");
    });

    it('should use displayName if defined & no "userDisplayName" exist', () => {
      const item = {
        displayName: "norris",
        name: "last",
      };
      const returnedSpan = render({ func: item });
      expect(
        returnedSpan
          .children()
          .first()
          .first()
          .text()
      ).toEqual("norris");
    });

    it('should use to name if no "userDisplayName"/"displayName" exist', () => {
      const item = {
        name: "last",
      };
      const returnedSpan = render({ func: item });
      expect(
        returnedSpan
          .children()
          .first()
          .first()
          .text()
      ).toEqual("last");
    });
  });

  describe("render parentheses", () => {
    let leftParen;
    let rightParen;

    beforeAll(() => {
      const item = { name: "" };
      const returnedSpan = render({ func: item });
      const children = returnedSpan.children();
      leftParen = returnedSpan.childAt(1);
      rightParen = returnedSpan.childAt(children.length - 1);
    });

    it("should be spans", () => {
      expect(leftParen.name()).toEqual("span");
      expect(rightParen.name()).toEqual("span");
    });

    it("should create a left paren", () => {
      expect(leftParen.text()).toEqual("(");
    });

    it("should create a right paren", () => {
      expect(rightParen.text()).toEqual(")");
    });
  });

  describe("render parameters", () => {
    let returnedSpan;
    let children;

    beforeAll(() => {
      const item = {
        name: "",
        parameterNames: ["one", "two", "three"],
      };
      returnedSpan = render({ func: item });
      children = returnedSpan.children();
    });

    it("should render spans according to the dynamic params given", () => {
      expect(children).toHaveLength(8);
    });

    it("should render the parameters names", () => {
      expect(returnedSpan.childAt(2).text()).toEqual("one");
    });

    it("should render the parameters commas", () => {
      expect(returnedSpan.childAt(3).text()).toEqual(", ");
    });
  });
});
