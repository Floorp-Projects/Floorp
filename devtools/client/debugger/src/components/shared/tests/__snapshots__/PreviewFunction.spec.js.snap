// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewFunction should return a span 1`] = `
<span
  className="function-signature"
>
  <span
    className="function-name"
  >
    &lt;anonymous&gt;
  </span>
  <span
    className="paren"
  >
    (
  </span>
  <span
    className="paren"
  >
    )
  </span>
</span>
`;
